torchao/prototype/scaled_grouped_mm/scaled_grouped_mm.py

```diff
@@ -0,0 +1,403 @@
+
+
+def _scaled_grouped_mm(
+    A: torch.Tensor,
+    B: torch.Tensor,
+    offs: torch.Tensor,
+    out_dtype: Optional[torch.dtype] = None,
+) -> torch.Tensor:
+    """
+    This function performs dynamic float8 quantization with row-wise scaling
+    on the input tensors A and B using the given recipe, then performs a
+    scaled grouped GEMM and returns the results.
+
+    Args:
+        A (bf16/float32 torch.Tensor): The first high-precision input tensor, which must be a 2D tensor of shape (M * num_groups, K).
+        B (bf16/float32 torch.Tensor): The second high-precision input tensor which must be 3D, which must be shape (B, K, N).
+        offs (int32 torch.Tensor): The offsets to use to mark the starting index of each group in the input tensor of shape.
+        float8_recipe (Float8LinearRecipeName): The recipe to use for dynamic float8 quantization.
+        out_dtype (Optional[torch.dtype]): The dtype of the output tensor. Currently only torch.bfloat16 is supported.
+        use_fast_accum (bool): Whether to use fast accumulation or not. Default is False.
+    """
+    return _Float8GroupedMM.apply(
+        A,
+        B,
+        offs,
+        out_dtype,
+    )
+
+
+class _Float8GroupedMM(torch.autograd.Function):
+    """Differentiable implementation of grouped GEMM with dynamic float8 quantization."""
+
+    @staticmethod
+    def forward(
+        ctx,
+        A: torch.Tensor,
+        B: torch.Tensor,
+        offs: torch.Tensor,
+        out_dtype: Optional[torch.dtype] = None,
+    ) -> torch.Tensor:
+        # torchao _scaled_grouped_mm only supports A=2D, B=3D.
+        assert A.ndim == 2, "A must be 2D"
+        assert B.ndim == 3, "B must be 3D"
+
+        assert (
+            A.size(-1) % 16 == 0
+        ), f"A must have a last dim divisible by 16, but got shape: {A.shape}"
+        assert (
+            B.size(-2) % 16 == 0 and B.size(-1) % 16 == 0
+        ), f"B must have last 2 dims divisible by 16, but got shape: {B.shape}"
+
+        # Assert input tensors are in high-precision dtypes.
+        assert (
+            A.dtype == torch.float32 or A.dtype == torch.bfloat16
+        ), "A must be float32 or bfloat16"
+        assert (
+            B.dtype == torch.float32 or B.dtype == torch.bfloat16
+        ), "B must be float32 or bfloat16"
+        assert offs.dtype == torch.int32, "offs must be int32"
+
+        # Assert A and B dims are compatible for a scaled grouped GEMM.
+        assert A.size(-1) == B.size(
+            -2
+        ), f"shape {A.shape} and {B.shape} are not compatible for _scaled_grouped_mm"
+
+        # Due to hardware requirements, the right operand in a scaled grouped GEMM must be column-major.
+        if not _is_column_major(B):
+            B_col_major = B.transpose(-2, -1).contiguous().transpose(-2, -1)
+        else:
+            B_col_major = B
+
+        # Fetch float8 config from specified recipe name.
+        float8_config = Float8LinearConfig.from_recipe_name(
```

Comment:
it's suprising to see a config created here, why not just inline the logic you need without worrying about configs?  IMO dealing with configs would be for when this API is about to be productionized.

Reply:
I agree, I used it here for this prototype because in the test code i need to use `matmul_with_hp_or_float8_args` (which requires a Float8 config) to compute the reference forward/backward - and trying to get the outputs/grads to match was already pretty tricky, so to start I wanted to just reference the same `Float8LinearConfig.from_recipe_name(Float8LinearRecipeName.ROWWISE)` everywhere, to minimize room for accidental differences in how a particular tensor is quantized in test code vs implementation, etc.

I am happy to refactor to inline everything in the implementation, but for the test code we still need the float8 config.